File: tests/test_expression_parser.py
from harpy import ExpressionParser, Lexer, SourceReader


class TestExpressionParser:
    def test_call(self):
        self._test("a()", "a()")
        self._test("a(b)", "a(b)")
        self._test("a(b, c)", "a(b, c)")
        self._test("a(b)(c)", "a(b)(c)")
        self._test("a(b) + c(d)", "(a(b) + c(d))")
        self._test("a(@b) + c(@d)", "(a((@b)) + c((@d)))")

    def test_unary_precedence(self):
        self._test("!-+a", "(!(-(+a)))")
        self._test("!!!a", "(!(!(!a)))")
        self._test("@a", "(@a)")

    def test_unary_binary_precedence(self):
        self._test("-a * b", "((-a) * b)")
        self._test("-a % b", "((-a) % b)")
        self._test("!a + b", "((!a) + b)")
        self._test("!a ^ b", "((!a) ^ b)")
        self._test("-a", "(-a)")
        self._test("!a", "(!a)")

    def test_binary_precedence(self):
        self._test(
            "a := b + c * d ^ e - f / g", "(a := ((b + (c * (d ^ e))) - (f / g)))"
        )

    def test_binary_associativity(self):
        self._test("a := b := c", "(a := (b := c))")
        self._test("a := b = c", "(a := (b = c))")
        self._test("a := b == c", "(a := (b == c))")
        self._test("a := b < c", "(a := (b < c))")
        self._test("a := b <= c", "(a := (b <= c))")
        self._test("a := b # c", "(a := (b # c))")
        self._test("a := b != c", "(a := (b != c))")
        self._test("a += b *= c", "(a += (b *= c))")
        self._test("a %= b -= c", "((a %= b) -= c)")
        self._test("a + b - c", "((a + b) - c)")
        self._test("a + b > c", "((a + b) > c)")
        self._test("a + b >= c", "((a + b) >= c)")
        self._test("a * b / c", "((a * b) / c)")
        self._test("a ^ b ^ c", "(a ^ (b ^ c))")
        self._test("a^b^c", "(a ^ (b ^ c))")
        self._test("a * b % c", "((a * b) % c)")
        self._test("a $ b $ c", "((a $ b) $ c)")
        self._test("a .or. b", "(a .or. b)")
        self._test("a .and. b", "(a .and. b)")
        self._test("a .and. b .or. c", "((a .and. b) .or. c)")
        self._test("a .or. b .and. c", "(a .or. (b .and. c))")
        self._test("!a .and. !b .or. !c", "(((!a) .and. (!b)) .or. (!c))")
        self._test("!(a .or. b .and. c)", "(!(a .or. (b .and. c)))")

    def test_grouping(self):
        self._test("a + (b + c) + d", "((a + (b + c)) + d)")
        self._test("a ^ (b + c)", "(a ^ (b + c))")
        self._test("(!a)", "(!a)")

    def test_literals(self):
        self._test(".t.", ".t.")
        self._test("123", "123")
        self._test("'hello'", "'hello'")
        self._test("NIL", "NIL")
        self._test("nil", "nil")

    def test_object_access(self):
        self._test("a:b", "a:b")
        self._test("a:b:c", "a:b:c")
        self._test("a:b()", "a:b()")
        self._test("a:b(c)", "a:b(c)")
        self._test("a:b():c", "a:b():c")
        self._test("a:b()[1]", "a:b()[1]")
        self._test("a:b()['key']", "a:b()['key']")
        self._test("a:b()[1 + c]", "a:b()[(1 + c)]")

    def test_array_declaration(self):
        self._test("a := { }", "(a := { })")
        self._test("a := { 1 }", "(a := { 1 })")
        self._test("a := { 1, 2 }", "(a := { 1, 2 })")
        self._test("a := { 1, b() }", "(a := { 1, b() })")
        self._test("a := { ;\n    1, ;\n    b() ;\n}", "(a := { 1, b() })")

    def test_hash_declaration(self):
        self._test("a := { => }", "(a := { => })")
        self._test("a := { 'b' => 1 }", "(a := { 'b' => 1 })")
        self._test("a := { 'b' => 1, 'c' => 2 }", "(a := { 'b' => 1, 'c' => 2 })")
        self._test("a := { 'b' => 1, 'c' => d() }", "(a := { 'b' => 1, 'c' => d() })")
        self._test(
            "a := { ;\n    'b' => 1, ;\n    'c' => d() ;\n}",
            "(a := { 'b' => 1, 'c' => d() })",
        )

    def test_codeblock(self):
        self._test("a := { || b() }", "(a := { || b() })")
        self._test("a := { |b| c(b) }", "(a := { |b| c(b) })")
        self._test("a := { |b,c| d(b, c) }", "(a := { |b, c| d(b, c) })")
        self._test("a := { |b,c| d(b), e(c) }", "(a := { |b, c| d(b), e(c) })")

    def test_conditional(self):
        self._test("a := iif(b, 1, 0)", "(a := iif(b, 1, 0))")

    def _test(self, source: str, expected: str):
        """Parses the given chunk of code and verifies that it matches the expected
        pretty-printed result.
        """
        lexer = Lexer(text=source)
        reader = SourceReader(tokens=lexer)
        parser = ExpressionParser(source_reader=reader)

        result = parser.parse()
        actual = result.print()

        assert actual == expected

File: HarpyTests/ParserTests/TestExpressionParser.cs
using Harpy.Lexer;
using Harpy.Parser;

namespace HarpyTests.ParserTests;

[TestClass]
public sealed class TestExpressionParser
{
    [TestMethod]
    public void TestCall()
    {
        AssertParsedEqualsExpected("a()", "a()");
        AssertParsedEqualsExpected("a(b)", "a(b)");
        AssertParsedEqualsExpected("a(b, c)", "a(b, c)");
        AssertParsedEqualsExpected("a(b)(c)", "a(b)(c)");
        AssertParsedEqualsExpected("a(b) + c(d)", "(a(b) + c(d))");
        AssertParsedEqualsExpected("a(@b) + c(@d)", "(a((@b)) + c((@d)))");
    }

    [TestMethod]
    public void TestUnaryPrecedence()
    {
        AssertParsedEqualsExpected("++a", "(++a)");
        AssertParsedEqualsExpected("++--a", "(++(--a))");
        AssertParsedEqualsExpected("++a--", "(++(a--))");
        AssertParsedEqualsExpected("!-+a", "(!(-(+a)))");
        AssertParsedEqualsExpected("!-++a", "(!(-(++a)))");
        AssertParsedEqualsExpected("!-a++", "(!(-(a++)))");
        AssertParsedEqualsExpected("!!!a", "(!(!(!a)))");
        AssertParsedEqualsExpected("@a", "(@a)");
    }

    [TestMethod]
    public void TestUnaryOrBinaryPrecedence()
    {
        AssertParsedEqualsExpected("-a * b", "((-a) * b)");
        AssertParsedEqualsExpected("-a % b", "((-a) % b)");
        AssertParsedEqualsExpected("!a + b", "((!a) + b)");
        AssertParsedEqualsExpected("!a ^ b", "((!a) ^ b)");
        AssertParsedEqualsExpected("-a", "(-a)");
        AssertParsedEqualsExpected("!a", "(!a)");
    }

    [TestMethod]
    public void TestBinaryPrecedence()
    {
        AssertParsedEqualsExpected(
            "a := b + c * d ^ e - f / g", "(a := ((b + (c * (d ^ e))) - (f / g)))"
        );
    }

    [TestMethod]
    public void TestBinaryAssociativity()
    {
        AssertParsedEqualsExpected("a := b := c", "(a := (b := c))");
        AssertParsedEqualsExpected("a := b = c", "(a := (b = c))");
        AssertParsedEqualsExpected("a := b == c", "(a := (b == c))");
        AssertParsedEqualsExpected("a := b < c", "(a := (b < c))");
        AssertParsedEqualsExpected("a := b <= c", "(a := (b <= c))");
        AssertParsedEqualsExpected("a := b # c", "(a := (b # c))");
        AssertParsedEqualsExpected("a := b != c", "(a := (b != c))");
        AssertParsedEqualsExpected("a += b *= c", "(a += (b *= c))");
        AssertParsedEqualsExpected("a %= b -= c", "((a %= b) -= c)");
        AssertParsedEqualsExpected("a + b - c", "((a + b) - c)");
        AssertParsedEqualsExpected("a + b > c", "((a + b) > c)");
        AssertParsedEqualsExpected("a + b >= c", "((a + b) >= c)");
        AssertParsedEqualsExpected("a * b / c", "((a * b) / c)");
        AssertParsedEqualsExpected("a ^ b ^ c", "(a ^ (b ^ c))");
        AssertParsedEqualsExpected("a^b^c", "(a ^ (b ^ c))");
        AssertParsedEqualsExpected("a * b % c", "((a * b) % c)");
        AssertParsedEqualsExpected("a $ b $ c", "((a $ b) $ c)");
        AssertParsedEqualsExpected("a .or. b", "(a .or. b)");
        AssertParsedEqualsExpected("a .and. b", "(a .and. b)");
        AssertParsedEqualsExpected("a .and. b .or. c", "((a .and. b) .or. c)");
        AssertParsedEqualsExpected("a .or. b .and. c", "(a .or. (b .and. c))");
        AssertParsedEqualsExpected("!a .and. !b .or. !c", "(((!a) .and. (!b)) .or. (!c))");
        AssertParsedEqualsExpected("!(a .or. b .and. c)", "(!(a .or. (b .and. c)))");
    }

    [TestMethod]
    public void TestGrouping()
    {
        AssertParsedEqualsExpected("a + (b + c) + d", "((a + (b + c)) + d)");
        AssertParsedEqualsExpected("a ^ (b + c)", "(a ^ (b + c))");
        AssertParsedEqualsExpected("(!a)", "(!a)");
    }

    [TestMethod]
    public void TestLiterals()
    {
        AssertParsedEqualsExpected(".t.", ".t.");
        AssertParsedEqualsExpected("123", "123");
        AssertParsedEqualsExpected("'hello'", "'hello'");
        AssertParsedEqualsExpected("NIL", "NIL");
        AssertParsedEqualsExpected("nil", "nil");
    }

    [TestMethod]
    public void TestObjectAccess()
    {
        AssertParsedEqualsExpected("a:b", "a:b");
        AssertParsedEqualsExpected("a:b:c", "a:b:c");
        AssertParsedEqualsExpected("a:b()", "a:b()");
        AssertParsedEqualsExpected("a:b(c)", "a:b(c)");
        AssertParsedEqualsExpected("a:b():c", "a:b():c");
        AssertParsedEqualsExpected("a:b()[1]", "a:b()[1]");
        AssertParsedEqualsExpected("a:b()['key']", "a:b()['key']");
        AssertParsedEqualsExpected("a:b()[1 + c]", "a:b()[(1 + c)]");
    }

    [TestMethod]
    public void TestArrayDeclaration()
    {
        AssertParsedEqualsExpected("a := { }", "(a := { })");
        AssertParsedEqualsExpected("a := { 1 }", "(a := { 1 })");
        AssertParsedEqualsExpected("a := { 1, 2 }", "(a := { 1, 2 })");
        AssertParsedEqualsExpected("a := { 1, b() }", "(a := { 1, b() })");
        AssertParsedEqualsExpected("a := { ;\n    1, ;\n    b() ;\n}", "(a := { 1, b() })");
    }

    [TestMethod]
    public void TestHashDeclaration()
    {
        AssertParsedEqualsExpected("a := { => }", "(a := { => })");
        AssertParsedEqualsExpected("a := { 'b' => 1 }", "(a := { 'b' => 1 })");
        AssertParsedEqualsExpected("a := { 'b' => 1, 'c' => 2 }", "(a := { 'b' => 1, 'c' => 2 })");
        AssertParsedEqualsExpected("a := { 'b' => 1, 'c' => d() }", "(a := { 'b' => 1, 'c' => d() })");
        AssertParsedEqualsExpected(
            "a := { ;\n    'b' => 1, ;\n    'c' => d() ;\n}",
            "(a := { 'b' => 1, 'c' => d() })"
        );
    }

    [TestMethod]
    public void TestCodeblock()
    {
        AssertParsedEqualsExpected("a := { || b() }", "(a := { || b() })");
        AssertParsedEqualsExpected("a := { |b| c(b) }", "(a := { |b| c(b) })");
        AssertParsedEqualsExpected("a := { |b,c| d(b, c) }", "(a := { |b, c| d(b, c) })");
        AssertParsedEqualsExpected("a := { |b,c| d(b), e(c) }", "(a := { |b, c| d(b), e(c) })");
    }

    [TestMethod]
    public void TestConditional()
    {
        AssertParsedEqualsExpected("a := iif(b, 1, 0)", "(a := iif(b, 1, 0))");
    }

    [TestMethod]
    public void TestComments()
    {
        AssertParsedEqualsExpected("a:b(c/* Not sure if this should be here */)", "a:b(c)");
        AssertParsedEqualsExpected("a:b(/* Not sure if this should be here */c)", "a:b(c)");
        AssertParsedEqualsExpected("a:b(c) // Comment at the end", "a:b(c)");
    }

    private static void AssertParsedEqualsExpected(string source, string expected)
    {
        var lexer = new Lexer(source);
        var reader = new SourceReader(lexer);
        var parser = new ExpressionParser(reader);

        var result = parser.Parse();
        var actual = result?.PrettyPrint();
        Assert.AreEqual(expected, actual);
    }
}